File: backend/internal/adapters/http/timezone.go
package httpapi

import (
	"net/http"
	"time"

	"flashcard-backend/internal/apperror"
)

// locationFromQuery reads the optional tz query parameter (an IANA time
// zone name such as "America/Sao_Paulo"). Endpoints that depend on what
// "today" is use it so days follow the user's own clock; without it days
// are UTC.
func locationFromQuery(r *http.Request) (*time.Location, error) {
	tz := r.URL.Query().Get("tz")
	if tz == "" {
		return time.UTC, nil
	}
	loc, err := time.LoadLocation(tz)
	if err != nil {
		return nil, apperror.Validation("timezone.invalid", "tz must be a valid IANA time zone name, e.g. America/Sao_Paulo")
	}
	return loc, nil
}

File: frontend/src/services/timezone.ts
// The browser's IANA time zone (e.g. "America/Manaus"). Sent to endpoints
// that depend on what "today" is, so the server counts the user's own days
// rather than UTC days.
export function browserTimeZone(): string | undefined {
  try {
    return Intl.DateTimeFormat().resolvedOptions().timeZone || undefined;
  } catch {
    return undefined;
  }
}

// "tz=America%2FManaus" (or an empty string when the zone is unknown), ready
// to be joined to other query parameters.
export function timeZoneParam(): string {
  const tz = browserTimeZone();
  return tz ? `tz=${encodeURIComponent(tz)}` : "";
}
